### PR TITLE
Upgrade hdf5 windows

### PR DESCRIPTION
**Pull request by LamarMoore**
**Description of work.**

Following an upgrade of `hdf5` for the windows platform [here](https://github.com/mantidproject/thirdparty-msvc2015/tree/upgrade-hdf5). The following changes were made to address compatibility issues with ESS nxs files.

**The changes to the thirdparty repo need to be merged before this PR is merged**

**To test:**

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/LoadEventNexus.cpp [hunk 1]
```diff
@@ -837,8 +848,13 @@ void LoadEventNexus::loadEvents(API::Progress *const prog,
          * the run_start from the proton_charge log. We are going to get this
          * from our event_time_zero instead
          */
-        auto localFirstLast = firstLastPulseTimes(*m_file, this->g_log);
-        firstPulseT = std::min(firstPulseT, localFirstLast.first);
+        try {
+          auto localFirst = firstPulseTime(*m_file, this->g_log);
+          firstPulseT = std::min(firstPulseT, localFirst);
+        } catch (std::runtime_error &) {
```
Comment (OwenArnold) [inline Framework/DataHandling/src/LoadEventNexus.cpp]: Why would this fail?

File: Framework/NexusGeometry/src/NexusGeometryParser.cpp [hunk 2]
```diff
@@ -822,6 +823,7 @@ class Parser {
       builder.addBank(bankName, bankPos, bankRotation);
       // Get the pixel detIds
       auto detectorIds = getDetectorIds(detectorGroup);
+      tbb::parallel_sort(detectorIds.begin(), detectorIds.end());
```
Comment (OwenArnold) [inline Framework/NexusGeometry/src/NexusGeometryParser.cpp]: Could probably use a comment about why you do this.

File: Framework/Types/src/Core/DateAndTime.cpp [hunk 3]
```diff
@@ -10,6 +10,8 @@ const uint32_t DateAndTime::EPOCH_DIFF = 631152000;
 /// The epoch for GPS times.
 const boost::posix_time::ptime
     DateAndTime::GPS_EPOCH(boost::gregorian::date(1990, 1, 1));
```
Comment (OwenArnold) [inline Framework/Types/src/Core/DateAndTime.cpp]: I know this wan't you, but I think the GPS_EPOCH above is wrong. GPS epoch 5th Jan 1980. Just a note, shouldn't really affect anything, but might cause confusion for future maintainers.

File: Framework/Kernel/inc/MantidKernel/NexusDescriptor.h [hunk 4]
```diff
@@ -90,6 +90,11 @@ class MANTID_KERNEL_DLL NexusDescriptor {
   std::string pathOfType(const std::string &type) const;
   /// Query if a given type exists somewhere in the file
   bool classTypeExists(const std::string &classType) const;
+  // Recursively search for the first group which contains a class of type
+  // classType
+  static bool findAndOpenParentGroup(::NeXus::File &file,
```
Comment (OwenArnold) [inline Framework/Kernel/inc/MantidKernel/NexusDescriptor.h]: The name of the function doesn't quite fit the description. Would imply opening the parent of some current with existing naming. `findAndOpenFirstChild`?

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Doesn't pinpoint issue, but a few changes suggested

**Comment by martyngigg** [comment]
While implementing the Python 3 builds for Windows I based my branch of just the HDF5 upgrade from the 3rd party. As far as I can tell the tests that fail here *do not* fail for me so I think the failing tests here are due to the other changes.

Could we possibly spin out the HDF5 upgrade only stuff to a separate PR?

**Comment by martyngigg** [comment]
Just as a note. I've started doing to the Python 3 work and I based the changes off the HDF5 1.10 upgrade in 3rd party but **not** with these changes. The above tests seem to pass there so I don't think it is an issue with the HDF5 libraries. Would it be possible to spin out just the HDF5 upgrade to a separate PR?

**Comment by martyngigg** [comment]
As discussed with Owen on Slack we decided that we should separate out the HDF5 upgrade and ESS compatibility changes as there are still issues with the latter. The HDF5 upgrade was merged in #27645 and #27646 now contains the other changes from this branch. The new branch is `ess-nexus-compat`.

Closing this as everything should now be on the aforementioned branch.

**Pull request closed without merging**